Plugins/ScriptingPython/scriptingpython.cpp

```diff
@@ -40,33 +67,196 @@ ScriptingPython::~ScriptingPython()
+#else
+    initPython();
+#endif
+    return true;
+}
+
+void ScriptingPython::initPython()
+{
+    QMutexLocker locker(mainInterpMutex);
+#ifdef PYTHON_DYNAMIC_BINDING
+    if (library.isLoaded())
+        deinitPython();
+    if (!bindLibrary())
+        return;
+    if (libraryVersion < QVersionNumber(3, 7))
+        PyImport_AppendInittab("db", pyDbModuleCompatInit);
+    else
+        PyImport_AppendInittab("db", pyDbModuleInit);
+#else
+#if PYTHON_VERSION_HEX < 0x03070000
+    PyImport_AppendInittab("db", &pyDbModuleCompatInit);
+#else
     PyImport_AppendInittab("db", &pyDbModuleInit);
+#endif
+#endif
+
     Py_Initialize();
+#ifdef PYTHON_DYNAMIC_BINDING
+    if (libraryVersion < QVersionNumber(3, 0))
+        PyUnicode_SetDefaultEncoding("utf-8");
+#endif
+
+    PyEval_InitThreads();  // No-op in Python 3.7+
+    // GIL is held, thread state is set.
+
     PyRun_SimpleString("import db");
 
     mainContext = new ContextPython();
     contexts[mainContext->interp] = mainContext;
-    return true;
 }
 
 void ScriptingPython::deinit()
 {
     QMutexLocker locker(mainInterpMutex);
+#ifdef PYTHON_DYNAMIC_BINDING
+    if (library.isLoaded())
+#endif
+        deinitPython();
+    SQLS_CLEANUP_RESOURCE(scriptingpython);
+}
+
+void ScriptingPython::deinitPython()
+{
     contexts.clear();
+
+    // Acquire GIL and set thread state
+    PyEval_RestoreThread(mainContext->interp);
     Py_Finalize();
-    SQLS_CLEANUP_RESOURCE(scriptingpython);
+    // GIL and thread state no longer exist
+
+    mainContext = nullptr;
+#ifdef PYTHON_DYNAMIC_BINDING
+    library.unload();
+#endif
 }
 
 QString ScriptingPython::getLanguage() const
 {
     return "Python";
 }
 
+#ifdef PYTHON_DYNAMIC_BINDING
+bool ScriptingPython::bindLibrary()
+{
+    library.setFileName(libraryPath);
+    if (!library.load())
+    {
+        qWarning() << "Could not load Python library" << libraryPath;
+        return false;
+    }
+
+    if (!bindSymbols(library))
+    {
+        qWarning() << "Could not bind all symbols in Python library: " << libraryPath;
+        library.unload();
+        return false;
+    }
+    libraryVersion = QVersionNumber::fromString(Py_GetVersion());
+    qInfo() << "Python library " << libraryPath << " has been loaded succesfully.";
+    return true;
+}
+
+void ScriptingPython::setLibraryPath(QString path)
+{
+    if (path != libraryPath)
+    {
+        libraryPath = path;
+        initPython();
+    }
+}
+
+static void globParts(QStringList& files, QList<QStringList> parts,
+                      QString prefix = QString(), int offset = 0)
+{
+    // Find all existing files among all concatenations of path parts.
+    // A part may include wildcards or path separators but not both.
+    if (parts.size() <= offset)
+        return;
+    bool isFile = offset == parts.size() - 1;
+    QDir::Filters filters = isFile ? QDir::Files : QDir::Dirs | QDir::NoDotAndDotDot | QDir::NoSymLinks;
+    for (QString piece : parts.at(offset))
```

Comment:
I think the "complexity" issue comes from here. It's a long, multiline body of the `for` loop, while it's not enclosed in braces. Especially that the body itself contains another multiline `if`s. It has potential for breaking the code more easily in future. Usually I allow myself to skip braces only if it's a single line body and if it's the last level of `if` or `loop`. So far did not get complaints about that.

(btw. it's the one of 2 reasons, for which I don't like Python - making indentation essential for code semantics.)

Reply:
Sure, I'll add the cosmetic braces. I've forgot about this piece.